### DISCUSSION TITLE
Update a global variable disregarding scope

### DISCUSSION THREAD

**Discussion by HitomiTenshi** [nushell/nushell#6417] [Q&A] [answered]
Hi, I have written a small update script where I want to output some text after it has run (or not run):

```sh
let update = false # <-- Declare variable

ls aur | each { |dir|
    cd $dir.name

    if not (git pull | str starts-with 'Already up to date.') {
        makepkg -sci
        git clean -xdf
        $update = true # <-- Assign value
    }
}

if $update { # <-- Test value
    echo 'Update completed successfully.'
} else {
    echo 'Already up to date.'
}
```

Running this script results in following error:

```
  × Parse mismatch during operation.
    ╭─[./update.nu:11:1]
 11 │         git clean -xdf
 12 │         $update = true
    ·                 ┬
    ·                 ╰── expected operator
 13 │     }
    ╰────
```

I looked through the documentation and have only found ways to shadow a variable by re-declaring it inside the scope [(e. g. as seen here)](https://www.nushell.sh/book/variables_and_subexpressions.html#variables), but what I want is to just update the global variable, disregarding the scope I'm currently in.

Is that possible?

I couldn't find basic variable operations / assignments in the documentation.

**Comment by fdncred** [comment, marked as answer]
Here's some docs that explain that variables are immutable. https://www.nushell.sh/book/thinking_in_nu.html#variables-are-immutable.

So, what you're trying to do isn't possible right now. However, we have discussed making some changes to bring mutability into nushell in some places.

**Comment by HitomiTenshi** [comment]
Thanks, now everything makes a lot more sense 😄

If anyone is interested in my updated solution, here is how I did it now using the `reduce` function:

```sh
let update = (
    ls aur | reduce -f false { |dir, acc| # <-- reduce with `false` as the initial value
        cd $dir.name

        if not (git pull | str starts-with 'Already up to date.') {
            makepkg -sci
            git clean -xdf
            true # <-- return `true`
        } else {
            $acc # <-- return the accumulated result (will be `true` if the above block was ever executed)
        }
    }
)

if $update {
    echo 'Update completed successfully.'
} else {
    echo 'Already up to date.'
}
```

You can execute further functions retaining the `$update` value by starting reduce with the previous result:

```sh
# First update:
let update = (
    ls aur | reduce -f false { |dir, acc| # <-- first run starts with `false`
        # your code ...
    }
)

# Consecutive updates:
let update = (
    ls other-folder | reduce -f $update { |dir, acc| # <-- Consecutive runs start with previous `$update` value
        # your code ...
    }
)
```

@fdncred As you can see this complicates things a lot and being able to have global state (and being able to update it inside scopes) would make things a lot easier.

**Reply by fdncred** [reply]
I agree. Working around immutability is hard for most people, including myself.
